### PR TITLE
Feat: Homepage

### PR DESCRIPTION
**Pull request by abdul-moize**
This branch adds the Homepage template to main branch

To test this PR. follow these steps

1. Pull branch and run: `python manage.py migrate && python manage.py runserver`. also create a super user by `python manage.py createsuper`.
2. Pull this branch & run: npm start
3. Open http://localhost:3000/ in browser

Browse the site and let me if there are any issues, comments & concerns

### CODE REVIEW

File: src/components/productRow/index.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import RowItem from "../rowItem";
+import "./productRow.css";
+
+function ProductRow(props) {
```
Comment (srehaider) [inline src/components/productRow/index.js:4]: Don't handle presentation using JS, you can handle it using CSS. Handling it like this means you will have 5 products on each row even on a smaller screen of a mobile. You should add all products in one container and then handle the number of products per row using CSS based on screen width.

File: src/services/productService.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import { PRODUCT_API } from "../constants";
+
+function NoProductsException(message) {
+  this.message = message;
+}
+
+export default function getProducts() {
+  return fetch(PRODUCT_API)
+    .then((res) => {
+      console.log(res);
```
Comment (srehaider) [inline src/services/productService.js:10]: Don't leave `console.log` in the final code.

File: src/services/productService.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      let row = [];
+
+      for (let i = 0; i < products.length; i++) {
+        row.push(products[i]);
+        if ((i + 1) % 5 === 0 || i === products.length - 1) {
+          newData.push(row);
+          row = [];
+        }
+      }
+      return newData;
```
Comment (srehaider) [inline src/services/productService.js:25]: No need to divide products into rows, it should be handled using CSS.

File: src/pages/homepage/index.js [hunk 4]
```diff
@@ -1,17 +1,19 @@
...
 function Homepage() {
   const [data, setData] = useState(null);
   const [errorMessage, setErrorMessage] = useState("");
   useEffect(() => {
     getProducts()
-      .then((newData) => setData(newData))
+      .then((newData) => {
+        console.log(newData);
+        setData(newData);
+      })
```
Comment (srehaider) [inline src/pages/homepage/index.js:15]:
```suggestion
      .then((products) => {
        setAvailableProducts(products);
      })
```

File: src/services/productService.js [hunk 5]
```diff
@@ -7,23 +7,10 @@ function NoProductsException(message) {
...
-      for (let i = 0; i < products.length; i++) {
-        row.push(products[i]);
-        if ((i + 1) % 5 === 0 || i === products.length - 1) {
-          newData.push(row);
-          row = [];
-        }
-      }
-      return newData;
-    })
+    .then((products) => products)
```
Comment (srehaider) [inline src/services/productService.js:13]: We don't need this `then`.
```suggestion
```

File: src/components/rowItem/index.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
+import { Link } from "react-router-dom";
+import { HOMEPAGE } from "../../constants";
+import "./rowItem.css";
+
+function RowItem(props) {
```
Comment (srehaider) [inline src/components/rowItem/index.js:5]: It's not a `RowItem` anymore it should be something like `ProductCard`.

File: src/pages/homepage/index.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import { useEffect, useState } from "react";
+import { ErrorField } from "../../components/elements";
+import getProducts from "../../services/productService";
+import "./homepage.css";
+import RowItem from "../../components/rowItem";
+
+function Homepage() {
+  const [data, setData] = useState(null);
```
Comment (srehaider) [inline src/pages/homepage/index.js:8]: We can give `data` state a more specific name like `availableProducts`.

File: src/pages/homepage/index.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+      });
+  }, []);
+  return (
+    <div className="main-container">
+      <img src="/homepage_banner.png" width="100%" alt="banner" />
+      <div className="products-container">
+        {data !== null ? (
+          data.map((element, index) => {
+            return <RowItem item={element} key={`row_${index}`} />;
+          })
```
Comment (srehaider) [inline src/pages/homepage/index.js:27]: Using index as a `key` prop is not recommended instead use `product.id`. Also try to use consistent names either use `product` or use `item` at all places, different names for same things cause confusion.

```suggestion
          availableProducts.map((product) => (
            <ProductCard product={product} key={product.id} />;
          ))
```

**Review by srehaider — CHANGES_REQUESTED** [review]
Few more improvements then we are good to merge.

File: src/pages/homepage/index.js (hunk 8 above)
Comment (abdul-moize) [reply src/pages/homepage/index.js]: Done

File: src/components/rowItem/index.js (hunk 6 above)
Comment (abdul-moize) [reply src/components/rowItem/index.js]: Done

File: src/services/productService.js (hunk 5 above)
Comment (abdul-moize) [reply src/services/productService.js]: Done

**Review by srehaider — APPROVED** [review]
Looks good to me :+1:

**Pull request merged**